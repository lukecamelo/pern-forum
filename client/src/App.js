import React, { Component } from 'react'
import styled from 'styled-components'

const Wrapper = styled.div`
  text-align: center;
`
const Header = styled.div`
  background-color: papayawhip;
  height: 150px;
  padding: 20px;
  color: palevioletred;
`

const Button = styled.button`
  background-color: palevioletred;
  padding: 15px 1em;
  color: papayawhip;
`

class App extends Component {
  state = {
    response: [],
    username: ''
  }

  componentDidMount = () => {
    this.apiCall()
      .then(response => {
        this.setState({ response })
      })
      .catch(err => console.log(err))
  }

  apiCall = async () => {
    const response = await fetch('/api/users', { method: 'GET' })
    const data = await response.json()

    if (response.status !== 200) throw Error(data.message)

    return data
  }

  addUserTest = () => {
    fetch('/api/threads', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ userId: 1, text: 'this is my thread content!' })
    })
  }

  render() {
    if (this.state.response.length) {
      return (
        <Wrapper>
          <Header>
            {this.state.response[0].username}
            <Button onClick={this.addUserTest}>click</Button>
          </Header>
        </Wrapper>
      )
    } else {
      return (
        <Header>
          <h1>nothing here</h1>
        </Header>
      )
    }
  }
}

export default App
