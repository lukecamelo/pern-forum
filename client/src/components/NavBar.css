.link {
  color: white;
  background-color: #0266c8;
  text-decoration: none;
  margin-right: 1em;
  padding: 14px;
  transition: .2s;
  border-radius: 5px;
}

.logout {
  color: #0266c8;
  background-color: white;
  text-decoration: none;
  margin-right: 1em;
  padding: 14px;
  transition: .2s;
  border-radius: 5px;
  border: 2px solid #0266c8;
}

.logout:hover {
  color: white;
  background-color: #f195ac;
  cursor: pointer;
}

.link:hover {
  color: white;
  background-color: #f195ac;
  cursor: pointer;
}