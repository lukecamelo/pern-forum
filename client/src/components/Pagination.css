.pagination-controls {
  display: flex;
  padding: 1em;
  justify-content: center;
}

.pagination-controls__button {
  background-color: white;
  /* border-radius: 0.15em; */
  text-decoration: none;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.1);
  cursor: default;
  margin: .5em .5em;
  padding: 0.5em 0.75em;
  box-shadow:  0 4px 8px 0 rgba(0,0,0,0.12),
  0 2px 4px 0 rgba(0,0,0,0.08);
  transition: .3s;
}

.pagination-controls__button:hover {
  cursor: pointer;
  transform: translateY(2px);
}

.pagination-controls__button--active {
  background-color: #4d7ea8;
  color: white;

  /* &:hover,
  &:hover:not(:active) {
    transform: none;
  } */
}
