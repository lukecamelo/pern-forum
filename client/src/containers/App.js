import React, { Component } from 'react'
import styled from 'styled-components'
import { Provider } from 'react-redux'
import store from '../store'

import Login from '../components/Login'
import Signup from '../components/Signup'

const Wrapper = styled.div`
  text-align: center;
`
const Header = styled.div`
  height: 150px;
  padding: 20px;
  color: #564154;
`

class App extends Component {
  state = {
    user: null,
    isLoggedIn: false
  }

  // componentDidMount = () => {
  //   if (localStorage.Authorization) {
  //     this.getUser()
  //   }
  // }

  mockLogin = () => {
    fetch('/auth/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ username: 'rediscover', password: 'testpassword' })
    })
      .then(res => res.json())
      .then(res => {
        localStorage.Authorization = res.token
        this.setState({
          isLoggedIn: true
        })
      })
  }

  getUser = () => {
    fetch('/user/profile', {
      method: 'GET',
      headers: {
        Authorization: 'Bearer ' + localStorage.Authorization
      }
    })
      .then(res => res.json())
      .then(res => {
        this.setState({
          user: res.username
        })
      })
  }

  render() {
    return (
      <Provider store={store}>
        <Wrapper>
          <Header>
            <Login />
            <Signup/>
          </Header>
        </Wrapper>
      </Provider>
    )
  }
}

export default App
