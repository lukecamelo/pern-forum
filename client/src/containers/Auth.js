import React from 'react'
import Login from '../components/Login'
import Signup from '../components/Signup'

import { connect } from 'react-redux'

const Auth = ({ isLoggedIn }) => {
  return <div>{isLoggedIn ? <Login /> : <Signup />}</div>
}

const mapStateToProps = state => ({
  isLoggedIn: state.auth.isLoggedIn
})

export default connect(mapStateToProps)(Auth)
