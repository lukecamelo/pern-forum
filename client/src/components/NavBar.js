import React from 'react'
import styled from 'styled-components'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { userLogout } from '../actions/authActions'
import { fetchThreads } from '../actions/threadActions'

import ResponsiveMenu from 'react-responsive-navbar'
import './NavBar.css'

const StyledNav = styled.nav`
  display: flex;
  justify-content: flex-end;
  padding: 15px;
  background-color: ${props => props.theme.primary};
  @media screen and (max-width: 700px) {
    flex-direction: column;
    align-items: center;
  }
`
const Logo = styled.div`
  display: flex;
  margin-right: auto;
  align-self: center;
  margin-left: 1em;
  color: white;
  font-size: 1.5rem;
  font-family: 'Lato', sans-serif;

  @media screen and (max-width: 701px) {
    display: none;
  }
`
const A = styled.a`
  color: white;
  background-color: #0266c8;
  text-decoration: none;
  margin-right: 1em;
  padding: 14px;
  transition: 0.2s;
  font-family: 'Lato', sans-serif;
  &:hover {
    color: white;
  background-color: #f195ac;
  cursor: pointer;
  }
`
export const NavBar = ({ isLoggedIn, user, isHome, ...props }) => {
  let navigation
  if (isLoggedIn) {
    navigation = (
      <StyledNav>
        <Logo>imp.zone</Logo>
        {!isHome ? (
          <Link to="/threads/1" className="navlink">
            Home
          </Link>
        ) : (
          <A className="navlink" onClick={() => props.fetchThreads()}>Home</A>
        )}
        <Link to="/usercontrolpanel" className="navlink">
          Control Panel
        </Link>
        <Link
          to="/threads/1"
          className="logout"
          onClick={() => props.userLogout()}
        >
          Logout ({user})
        </Link>
      </StyledNav>
    )
    return (
      <ResponsiveMenu
        menuOpenButton={<i className="fas fa-bars fa-3x" />}
        menuCloseButton={<i className="fas fa-times-circle" />}
        changeMenuOn="700px"
        menu={navigation}
        largeMenuClassName="large-menu"
        smallMenuClassName="small-menu"
      />
    )
  } else {
    navigation = (
      <StyledNav>
        <Logo>imp.zone</Logo>
        <Link to="/threads/1" className="navlink">
          Home
        </Link>

        <Link to="/login" className="navlink">
          Login
        </Link>

        <Link to="/signup" className="navlink">
          Signup
        </Link>
      </StyledNav>
    )
    return (
      <ResponsiveMenu
        menuOpenButton={<i className="fas fa-bars fa-3x" />}
        menuCloseButton={<i className="fas fa-times-circle" />}
        changeMenuOn="700px"
        menu={navigation}
        largeMenuClassName="large-menu"
        smallMenuClassName="small-menu"
      />
    )
  }
}

const mapStateToProps = state => ({
  isLoggedIn: state.auth.isLoggedIn,
  user: state.auth.username
})

export default connect(
  mapStateToProps,
  { userLogout, fetchThreads }
)(NavBar)
