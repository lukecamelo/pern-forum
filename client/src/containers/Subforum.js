import React, { Component } from 'react'
import { connect } from 'react-redux'
import { fetchSubforumThreads, fetchData } from '../actions/threadActions'
import ThreadList from './ThreadList'
import Pagination from '../components/Pagination'
import { Banner } from './App'
import { H1, SubforumLink } from '../styled'
import NavBar from '../components/NavBar'
import Loader from '../components/Loader'

class Subforum extends Component {
  state = {
    hasLoaded: false,
    subforumName: ''
  }

  async componentDidMount() {
    await this.props.fetchData()
    switch (this.props.match.params.id) {
      case '1':
        this.setState({ subforumName: 'General Discussion' })
        break
      case '2':
        this.setState({ subforumName: 'Video Games' })
        break
      default:
        break
    }
    this.setState({
      hasLoaded: true
    })
  }

  render() {
    if (this.state.hasLoaded) {
      return (
        <React.Fragment>
          <NavBar />
          <Banner>
            <div>
              <SubforumLink
                to="/subforums"
                style={{ margin: '0 auto', color: 'white' }}
              >
                Forums
              </SubforumLink>{' '}
              <H1 style={{ margin: '0 auto', color: 'white', display: 'inline' }}> -> </H1>{' '}
              <H1 style={{ margin: '0 auto', color: 'white', display: 'inline' }}>
                {this.state.subforumName}
              </H1>
            </div>
          </Banner>
          <Pagination
            data={this.props.threads}
            currentPage={this.props.match.params.page}
            context="threads"
            pageSize={15}
            isLoggedIn={this.props.auth.isLoggedIn}
            subforumId={this.props.match.params.id}
          >
            {data => (
              <ThreadList data={data} subforumId={this.props.match.params.id} />
            )}
          </Pagination>
        </React.Fragment>
      )
    } else {
      return <Loader />
    }
  }
}

const mapStateToProps = state => ({
  threads: state.threadData.threads,
  auth: state.auth
})

export default connect(
  mapStateToProps,
  { fetchSubforumThreads, fetchData }
)(Subforum)
