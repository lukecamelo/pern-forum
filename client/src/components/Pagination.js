import React, { Component } from 'react'
import * as Scroll from 'react-scroll'
import { Link } from 'react-router-dom'
import { NewThreadLink } from '../containers/App'
import './Pagination.css'

class Pagination extends Component {
  state = {
    currentPage: null,
    pageCount: null
  }

  componentDidMount = () => {
    const startingPage = this.props.startingPage ? this.props.startingPage : 1
    const data = this.props.data
    const pageSize = this.props.pageSize
    let pageCount = parseInt(data.length / pageSize, 10)
    if (data.length % pageSize > 0) {
      pageCount++
    }
    this.setState({
      currentPage: startingPage,
      pageCount: pageCount
    })
  }

  setCurrentPage = num => {
    let scroll = Scroll.animateScroll
    if (this.props.context === 'threads') {
      scroll.scrollToTop({ duration: 300, smooth: 'easeInOutQuint' })
    } else {
      scroll.scrollToTop({ duration: 500, smooth: 'easeInOutQuint' })
    }
    this.setState({ currentPage: num })
  }

  createControls = () => {
    let controls = []
    const pageCount = this.state.pageCount
    for (let i = 1; i <= pageCount; i++) {
      const baseClassName = 'pagination-controls__button'
      const activeClassName =
        i === parseInt(this.props.currentPage, 10)
          ? `${baseClassName}--active`
          : ''
      controls.push(
        <Link
          to={
            this.props.context === 'threads'
              ? `/threads/${i}`
              : `/thread/${this.props.threadId}/page/${i}`
          }
          className={`${baseClassName} ${activeClassName}`}
          key={i}
          onClick={() => this.setCurrentPage(i)}
        >
          {i}
        </Link>
      )
    }
    return controls
  }

  createPaginatedData = () => {
    const { data, pageSize } = this.props
    const currentPage = this.props.currentPage
    const upperLimit = currentPage * pageSize
    const dataSlice = data.slice(upperLimit - pageSize, upperLimit)
    return dataSlice
  }

  render() {
    const { context } = this.props
    return (
      <div className="pagination">
        <div className="pagination-results">
          {this.props.children(this.createPaginatedData())}
        </div>
        <div className={context === 'threads' ? 'bottom' : 'posts-bottom'}>
          {this.state.pageCount > 1 ? (
            <div className="pagination-controls">
              <span>{this.createControls()}</span>
            </div>
          ) : null}
          {context === 'threads' ? (
            <div style={{ margin: '1em 0 2em 0', textAlign: 'center' }}>
              <NewThreadLink to="/newthread">Post Thread</NewThreadLink>
            </div>
          ) : null}
        </div>
      </div>
    )
  }
}

Pagination.defaultProps = {
  pageSize: 20,
  startingPage: 1
}

export default Pagination
