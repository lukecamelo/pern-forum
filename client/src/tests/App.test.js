import React from 'react'
import PropTypes from 'prop-types'
import { App } from '../containers/App'
import { shallow } from 'enzyme'
import NavBar from '../components/NavBar'
import Pagination from '../components/Pagination'

import renderAppWithState from './testUtils/utils'

describe('renders without crashing', () => {
  let wrapper, props
  beforeEach(() => {
    props = {
      threads: [
        { title: 'hello', content: 'hello hello', userId: 1, id: 1 },
        { title: 'hello', content: 'hello hello', userId: 1, id: 1 },
        { title: 'hello', content: 'hello hello', userId: 1, id: 1 }
      ],
      fetchThreads: () => [],
      fetchData: () => [],
      checkUserLoggedIn: () => null,
      isLoggedIn: true,
      match: {
        params: {
          page: 1
        }
      }
    }
    wrapper = shallow(<App {...props} />, {
      context: {
        router: {
          history: {
            push: () => {},
            replace: () => {},
            createHref: () => {}
          }
        }
      },
      childContextTypes: {
        router: PropTypes.object.isRequired
      }
    })
  })

  it('renders children', () => {
    expect(wrapper.find(Pagination).length).toEqual(1)
    expect(wrapper.find(NavBar).length).toEqual(1)
  })

  it('only renders posts if isLoggedIn equals true', () => {
    expect(wrapper.find(Pagination).length).toEqual(1)
    wrapper.setProps({ isLoggedIn: false })
    expect(wrapper.find(Pagination).length).toEqual(0)
  })

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

})
