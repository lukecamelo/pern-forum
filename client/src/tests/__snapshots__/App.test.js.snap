// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    checkUserLoggedIn={[Function]}
    fetchData={[Function]}
    fetchThreads={[Function]}
    isLoggedIn={true}
    match={
        Object {
            "params": Object {
              "page": 1,
            },
          }
    }
    threads={
        Array [
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Connect(NavBar) />,
        <Pagination
          context="threads"
          currentPage={1}
          data={
                    Array [
                              Object {
                                "content": "hello hello",
                                "id": 1,
                                "title": "hello",
                                "userId": 1,
                              },
                              Object {
                                "content": "hello hello",
                                "id": 1,
                                "title": "hello",
                                "userId": 1,
                              },
                              Object {
                                "content": "hello hello",
                                "id": 1,
                                "title": "hello",
                                "userId": 1,
                              },
                            ]
          }
          pageSize={10}
          startingPage={1}
>
          [Function]
</Pagination>,
        <Link
          className="thread-button"
          replace={false}
          to="/newthread"
>
          Post Thread
</Link>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": [Function],
          "context": "threads",
          "currentPage": 1,
          "data": Array [
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
            Object {
              "content": "hello hello",
              "id": 1,
              "title": "hello",
              "userId": 1,
            },
          ],
          "pageSize": 10,
          "startingPage": 1,
        },
        "ref": null,
        "rendered": [Function],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Post Thread",
          "className": "thread-button",
          "replace": false,
          "to": "/newthread",
        },
        "ref": null,
        "rendered": "Post Thread",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Connect(NavBar) />,
          <Pagination
            context="threads"
            currentPage={1}
            data={
                        Array [
                                    Object {
                                      "content": "hello hello",
                                      "id": 1,
                                      "title": "hello",
                                      "userId": 1,
                                    },
                                    Object {
                                      "content": "hello hello",
                                      "id": 1,
                                      "title": "hello",
                                      "userId": 1,
                                    },
                                    Object {
                                      "content": "hello hello",
                                      "id": 1,
                                      "title": "hello",
                                      "userId": 1,
                                    },
                                  ]
            }
            pageSize={10}
            startingPage={1}
>
            [Function]
</Pagination>,
          <Link
            className="thread-button"
            replace={false}
            to="/newthread"
>
            Post Thread
</Link>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": [Function],
            "context": "threads",
            "currentPage": 1,
            "data": Array [
              Object {
                "content": "hello hello",
                "id": 1,
                "title": "hello",
                "userId": 1,
              },
              Object {
                "content": "hello hello",
                "id": 1,
                "title": "hello",
                "userId": 1,
              },
              Object {
                "content": "hello hello",
                "id": 1,
                "title": "hello",
                "userId": 1,
              },
            ],
            "pageSize": 10,
            "startingPage": 1,
          },
          "ref": null,
          "rendered": [Function],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Post Thread",
            "className": "thread-button",
            "replace": false,
            "to": "/newthread",
          },
          "ref": null,
          "rendered": "Post Thread",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "childContextTypes": Object {
      "router": [Function],
    },
    "context": Object {
      "router": Object {
        "history": Object {
          "createHref": [Function],
          "push": [Function],
          "replace": [Function],
        },
      },
    },
    "hydrateIn": undefined,
  },
}
`;
